Rework appliances

trasher opened the pull request:
Appliance feature was using code that has been copy/pasted from the original plugin; but since license was not compatible, this should not have happen.

First commit of the current pull request first removes all appliances plugin code that may have been copied; and second is a brand new implementation, "from scratch" (and other classes already present in GLPI :p).

For now, the ApplianceRelation feature has not been reintegrated. I've not been able to understand completely the feature and/or there are implementation issues in existing code. Tables and fields will be kept until a solution has been found.

As far as I understand, we should be able to link several domains, networks or locations on `ApplianceItem`; but this is currently limited to only one of those. I'm not sure we already have such a working case (this means applying a `CommonDBRelation` on 2 other `CommonDBRelation`), and therefore if that is handled.

Plus, a new `is_helpdesk_visible` field has been added in the appliances table.

orthagh commented:
@yllen if you have time to discuss about the PR feel free.

yllen commented on front/appliance.form.php:
```diff
@@ -0,0 +1,104 @@
+              //TRANS: %s is the user login
+              sprintf(__('%s updates an item'), $_SESSION["glpiname"]));
+   Html::back();
+
```
Missing dellieu and addlieu

trasher replied:
Those are specific to the plugin; and seems related to ApplianceRelation problem described in PR description.

yllen replied:
we can't use the Appliances without the relations (explanation done by mail).

trasher replied:
Well, what your are talking about is to setup a CommonDBRelation on two other CommonDBRelation. As said, I'm not sure it can work; and it will certainly be a nightmare from an UI point of view.
We'll make tests and work on that specific point on another branch; but it certainly won't be finished for the 9.5.2 release.

yllen commented on inc/appliance.class.php:
```diff
@@ -0,0 +1,436 @@
+      echo "</td>";
+      $randDropdown = mt_rand();
+      echo "<td><label for='dropdown_manufacturers_id$randDropdown'>".__('Manufacturer')."</label></td>";
+      echo "<td>";
```
an Appliance is a group of different items with there own manufacturer.
So this field is unneeded

trasher replied:
Appliance feature in GLPI is a mix beetween Appliances and Webapplications plugins; this field was present in Webapplications. Also, it seems it should make sense on some cases, so since this is not a mandatory field, it can stay as is.

yllen commented on inc/appliance.class.php:
```diff
@@ -0,0 +1,436 @@
+      echo "<td>" . __('Associable to a ticket') . "</td><td>";
+      Dropdown::showYesNo('is_helpdesk_visible', $this->fields['is_helpdesk_visible']);
+      echo "</td></tr>\n";
+
```
Missing Item to link and Comments

trasher replied:
Comments has been added on form. As for adding an item, it seems this is also related to ApplianceRelation.

yllen commented on inc/appliance_item.class.php:
```diff
@@ -0,0 +1,403 @@
+
+   static public $itemtype_1 = 'Appliance';
+   static public $items_id_1 = 'appliances_id';
+
```
missing $take_entity_1 and $take_entity_2 (not default values)

Review not finished

yllen commented:
Review not finished (I have to urgently troubleshoot a customert)

trasher commented:
So, Relation has been added back to Appliances, with schema changes to add the itemtype and rename table.
This will be part of the 9.5.2 release coming soon.

cedric-anne commented:
Adding relations in the `Items` tab of an `Appliance` is simple and clear :+1: .
It may be interessing to have the same feature in the `Appliances` tab of an asset.

cedric-anne commented on inc/appliance.class.php:
```diff
@@ -54,543 +60,384 @@ static function getTypeName($nb = 0) {
-      }
-      echo "</td></tr>";
+
+      $rand = mt_rand();
```
Already initialized.
```suggestion
```

cedric-anne commented on inc/appliance_item.class.php:
```diff
@@ -29,417 +33,352 @@
-      );
+      switch ($item->getType()) {
+         case Appliance::class:
+            self::showItems($item, $withtemplate);
```
```suggestion
            self::showItems($item);
```

orthagh commented:
> Adding relations in the `Items` tab of an `Appliance` is simple and clear 👍 .
> It may be interessing to have the same feature in the `Appliances` tab of an asset.

Done partially, i'm not very fan of the double function (`<form>` tag must be separated from the list, as we already in a massiveaction form) but i don't have anymore idea.

In the afternoon, if the first part ok, i'll add some dockblock (ping @cedric-anne)

yllen commented:
I try to do a migration from appliances to use these new tables, but is not easy

trasher commented:
> I try to do a migration from appliances to use these new tables, but is not easy

I'll take an eye to add new changes in https://github.com/glpi-project/glpi/pull/7717 tomorrow; I've not reviewed it yet, but this looks to be a good start.

orthagh commented:
> a same appliance can't have different type of relations (only one type by appliance).
> Many relations for this type but only one type

Why not permits multiple type of relations ? 
It doesn't change anything from database point of view.
I can think about an appliance where an item have a linked domain and a location.
Maybe you can explain why the limitation is important

> display type of relation (actually only the name is displayed)
> display serial number and inventory number

Done:
![image](https://user-images.githubusercontent.com/418844/94785552-1bd5a800-03d0-11eb-81f4-a7edd1a7f758.png)

yllen commented:
If you permit multiple relations for a same appliance, you change all the functionality of software based on this appliance (i think about CNAMTS).
I think this functionality can adding in a second version to leave time of use to change their own code.

About migration, this one must be done in the plugin because it must still exist for added fields.

trasher commented on inc/console/migration/appliancesplugintocorecommand.class.php:
```diff
@@ -0,0 +1,743 @@
+      );
+
+      $table  = Infocom::getTable();
+      $result = $this->db->delete($table, [
+         'itemtype' => self::CORE_APPLIANCE_ITEMTYPE
+      ]);
```
All existing data from core appliances are dropped at the start of the migration procedure, so all infocom on core appliance type are deleted

trasher commented on inc/console/migration/appliancesplugintocorecommand.class.php:
```diff
@@ -0,0 +1,743 @@
+      return 1;
+   }
+
+   private function updateItemtypes() {
```
So not, we cannot use that. Data from the plugin must stay as is

cedric-anne commented on inc/console/migration/appliancesplugintocorecommand.class.php:
```diff
@@ -0,0 +1,743 @@
+            '<error>' . $message . '</error>',
+            OutputInterface::VERBOSITY_QUIET
+         );
+         return self::ERROR_PLUGIN_IMPORT_FAILED;
```
Done.

cedric-anne commented:
Search option "associated items" may be improved 
![image](https://user-images.githubusercontent.com/33253653/95053780-315d1180-06f1-11eb-8bf3-f38a6ab53bdb.png)

cedric-anne commented:
> Search option "associated items" may be improved
> ![image](https://user-images.githubusercontent.com/33253653/95053780-315d1180-06f1-11eb-8bf3-f38a6ab53bdb.png)

Fixed by 24a5bfe58ec8774889ec9c7e7a720d58ef08a13e

trasher commented:
I've rebased the branch and squashed commits; to prepare merge of this pull request today; since most of the job has been done and must be part of 9.5.2.
